<!DOCTYPE HTML>
<html lang="de">
<head>
<title>OnError - Syntax &amp; Verwendung | AutoHotkey v2</title>
<meta name="description" content="Die OnError-Funktion registriert eine Funktion, die jedes Mal automatisch aufgerufen wird, wenn ein unbehandelter Fehler auftritt." />
<meta http-equiv="content-type" content="text/html; charset=UTF-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<link href="../static/theme.css" rel="stylesheet" type="text/css" />
<script src="../static/content.js" type="text/javascript"></script>
<script src="../static/ga4.js" type="text/javascript"></script>
</head>
<body>

<h1>OnError</h1>
<p>Registriert eine <a href="../Functions.htm">Funktion</a>, die jedes Mal automatisch aufgerufen wird, wenn ein unbehandelter Fehler auftritt.</p>

<pre class="Syntax"><span class="func">OnError</span> Rückruf <span class="optional">, HinzufügenEntfernen</span></pre>

<h2 id="Parameters">Parameter</h2>
<dl>

  <dt>Rückruf</dt>
  <dd>
    <p>Typ: <a href="../misc/Functor.htm">Funktionsobjekt</a></p>
    <p>Eine Funktion, die aufgerufen werden soll.</p>
    <p>Die Rückruffunktion akzeptiert zwei Parameter und kann wie folgt <a href="../Functions.htm#intro">definiert</a> werden:</p>
    <pre class="NoIndent">MeinRückruf(ThrowWert, Modus) { ...</pre>
    <p>Es spielt keine Rolle, welche Namen Sie den Parametern geben, allerdings werden ihnen die folgenden Werte nacheinander zugewiesen:</p>
    <ol>
      <li>Der geworfene Wert, was üblicherweise ein <a href="Error.htm">Error-Objekt</a> ist.</li>
      <li>Der Fehlermodus: Return, Exit oder ExitApp. Einzelheiten finden Sie in der <a href="#Error_Modes">unteren Tabelle</a>.</li>
    </ol>
    <p>Es können beliebig viele Parameter am Ende der Parameterliste der Rückruffunktion weggelassen werden, wenn die entsprechenden Informationen nicht benötigt werden, aber dann muss als letzter Parameter ein Sternchen angegeben werden, z.B. <code>MeinRückruf(Param1, *)</code>.</p>
    <p>Die Rückruffunktion kann einen der folgenden Werte zurückgeben (andere Werte sind für zukünftige Zwecke reserviert und sollten vermieden werden):</p>
    <ul>
      <li><code>0</code>, <code>""</code> oder kein Return: Fehlerbehandlung normal fortsetzen lassen.</li>
      <li><code>1</code>: Das reguläre Fehlerdialogfenster und alle verbleibenden Rückruffunktionen unterdrücken.</li>
      <li><code>-1</code>: Wie oben, aber wenn <em>Modus</em> (der zweite Parameter) das Wort Return enthält, darf die Ausführung des aktuellen Threads fortgesetzt werden.</li>
    </ul>
  </dd>
  
  <dt>HinzufügenEntfernen</dt>
  <dd>
    <p>Typ: <a href="../Concepts.htm#numbers">Integer</a></p>
    <p>Wenn weggelassen, wird standardmäßig 1 verwendet. Andernfalls geben Sie eine der folgenden Zahlen an:</p>
    <ul>
      <li>1 = Die Rückruffunktion nach anderen zuvor registrierten Rückruffunktionen aufrufen.</li>
      <li>-1 = Die Rückruffunktion vor anderen zuvor registrierten Rückruffunktionen aufrufen.</li>
      <li>0 = Die Rückruffunktion nicht aufrufen.</li>
    </ul>
  </dd>

</dl>

<h2 id="Error_Modes">Fehlermodi</h2>
<table class="info">
  <tr>
    <th>Modus</th>
    <th abbr="Beschr">Beschreibung</th>
  </tr>
  <tr>
    <td>Return</td>
    <td>Der geworfene Wert ist ein fortsetzbarer Laufzeitfehler. Der Thread wird fortgesetzt, wenn die Rückruffunktion -1 zurückgibt, andernfalls wird der Thread beendet.</td>
  </tr>
  <tr>
    <td>Exit</td>
    <td>Der geworfene Wert ist ein nicht-fortsetzbarer Laufzeitfehler oder ein vom Skript <a href="Throw.htm">geworfener</a> Wert. Der Thread wird beendet.</td>
  </tr>
  <tr>
    <td>ExitApp</td>
    <td>Der geworfene Wert ist ein kritischer Laufzeitfehler, z.B. eine von DllCall erkannte Datenbeschädigung. Das Programm wird beendet.</td>
  </tr>
</table>

<h2 id="Remarks">Bemerkungen</h2>
<p><em>Rückruf</em> wird nur bei Fehlern oder Ausnahmen aufgerufen, die normalerweise zu einer Fehlermeldung führen würden. Ladezeitfehler gehören nicht dazu, da OnError erst nach dem Laden des Skripts aufgerufen werden kann.</p>
<p><em>Rückruf</em> wird im aktuellen <a href="../misc/Threads.htm">Thread</a> vor seiner Beendigung aufgerufen (also bevor der Aufrufstapel abgearbeitet wurde).</p>

<h2 id="Related">Siehe auch</h2>
<p><a href="Try.htm">Try</a>, <a href="Catch.htm">Catch</a>, <a href="Throw.htm">Throw</a>, <a href="OnExit.htm">OnExit</a></p>

<h2 id="Examples">Beispiele</h2>
<div class="ex" id="ExBasic">
<p><a class="ex_number" href="#ExBasic"></a> Protokolliert vom Skript verursachte Fehler in einer Textdatei, anstatt sie dem Benutzer anzuzeigen.</p>
<pre>
OnError FehlerProtokollieren
i := Integer("Fehler_verursachen")

FehlerProtokollieren(Exception, Modus) {
    FileAppend "Fehler bei Zeile " Exception.Line ": " Exception.Message "`n"
        , "Fehlerprotokoll.txt"
    return true
}</pre>
</div>

<div class="ex" id="ExAccumulator">
<p><a class="ex_number" href="#ExAccumulator"></a> Verwenden Sie OnError, um alternative Fehlerbehandlungsmethoden zu implementieren. Einschränkung: OnError ist unwirksam, während <a href="Try.htm">Try</a> aktiv ist.</p>
<pre>
MehrereFehlerAkkumulieren()
{
    local ea := FehlerAkkumulator()
    ea.Start()
    return ea
}

class FehlerAkkumulator
{
    Errors := []                        <em>; Array für akkumulierte Fehler.</em>
    _cb := FehlerAkkumulieren.Bind(this.Errors)
    Start() =&gt; OnError(this._cb, -1)    <em>; Unsere cb vor allen anderen registrieren.</em>
    Stop() =&gt; OnError(this._cb, 0)      <em>; Unsere cb deregistrieren.</em>
    Last =&gt; this.Errors[-1]             <em>; Aktuellster Fehler.</em>
    Count =&gt; this.Errors.Length         <em>; Anzahl der akkumulierten Fehler.</em>
    __item[i] =&gt; this.Errors[i]         <em>; Shortcut für Indexierung.</em>
    __delete() =&gt; this.Stop()           <em>; Zum Binden an Gültigkeitsbereich der Funktion.</em>
}

<em>; Nachfolgend die OnError-Rückruffunktion. 'errors' bekommt einen Wert via Bind() zugewiesen.</em>
FehlerAkkumulieren(errors, e, mode)
{
    if mode != "Return" <em>; Nicht fortsetzbar.</em>
        return
    if e.What = "" <em>; Ausdrucksfehler oder ähnliches, keine interne Funktion.</em>
        return
    try {
        <em>; Versuchen, den Fehler in stdout auszugeben.</em>
        FileAppend Format("{1} ({2}) : ({3}) {4}`n", e.File, e.Line, e.What, e.Message), "*"
        if HasProp(e, "extra")
            FileAppend "     Specifically: " e.Extra "`n", "*"
    }
    errors.Push(e)
    return -1 <em>; Weiter.</em>
}

FensterNeuanordnen()
{
    <em>; Beginnen, Fehler in 'err' zu akkumulieren.</em>
    local err := MehrereFehlerAkkumulieren()

    <em>; Einige Dinge tun, die fehlschlagen können...</em>
    MonitorGetWorkArea , &amp;Links, &amp;Oben, &amp;Rechts, &amp;Unten
    Breite := (Rechts-Links)//2, Höhe := Unten-Oben
    WinMove Links, Oben, Breite, Höhe, A_ScriptFullPath
    WinMove Links+Breite, Oben, Breite, Höhe, "Hilfe für AutoHotkey v2"

    <em>; Prüfen, ob Fehler aufgetreten sind.</em>
    if err.Count
        MsgBox err.Count " Fehler; letzter Fehler auf Zeile Nr. " err.Last.Line
    else
        MsgBox "Keine Fehler"

    <em>; Stop wird automatisch aufgerufen, wenn die Variable den Gültigkeitsbereich
    ; verlässt, da nur wir eine Referenz zum Objekt haben. Dies führt dazu,
    ; dass OnError aufgerufen wird, um die Rückruffunktion zu deregistrieren.
    ;err.Stop()</em>
}

<em>; Testfunktion aufrufen, die Fehler unterdrückt und akkumuliert.</em>
FensterNeuanordnen()
<em>; Andere Funktion aufrufen, um zu zeigen, dass normales Fehlerverhalten wiederhergestellt ist.</em>
WinMove 0, 0, 0, 0, "Nicht-existierendes Fenster"
</pre>
</div>

</body>
</html>
